environment.py

```diff
@@ -0,0 +1,422 @@
+import pybullet as p
+import pybullet_data
+import time
+import math
+from datetime import datetime
+import numpy as np
+import time
+from scipy.spatial.transform import Rotation as R
+import scipy.stats
+import typing as T
+
+from sim_objects import Arm, Bottle
+
+
+class ActionSpace():
+    """Action space defined by incremental changes to individual joints.
+    These include positive and negative offsets and no-change if specified
+    """
+    default_da_rad = 5.0 * math.pi / 180.0  # default 5 degrees offsets
+
+    def __init__(self, num_DOF, da_rad=default_da_rad, include_no_change=False,
+                 ignore_last_joint=True):
+        self.num_DOF = num_DOF
+        # to ensure each action's new state is treated as new state due to discretization
+        self.da_rad = da_rad * 1.2
+        # self.traj_iter_set = [100, 150, 200]
+        self.traj_iter_set = [200]
+
+        pos_moves = np.eye(N=num_DOF) * self.da_rad
+        neg_moves = np.eye(N=num_DOF) * -self.da_rad
+        no_change = np.zeros((1, num_DOF))
+
+        if ignore_last_joint:
+            pos_moves = pos_moves[:-1, :]
+            neg_moves = neg_moves[:-1, :]
+
+        if include_no_change:
+            self.actions_mat = np.vstack([
+                no_change, pos_moves, neg_moves
+            ])
+        else:
+            self.actions_mat = np.vstack([
+                pos_moves, neg_moves
+            ])
+        self.num_actions = self.actions_mat.shape[0] * len(self.traj_iter_set)
+        self.action_ids = list(range(self.num_actions))
+
+    def get_action(self, id):
+        assert(isinstance(id, int))
+        assert (0 <= id < self.num_actions)
+        dq_i = id % self.actions_mat.shape[0]
+        num_iters = self.traj_iter_set[int(id / self.actions_mat.shape[0])]
+        return (self.actions_mat[dq_i, :], num_iters)
+
+
+class EnvParams(object):
+    def __init__(self, bottle_fill, bottle_fric, bottle_fill_prob,
+                 bottle_fric_prob):
+        self.bottle_fill = bottle_fill
+        self.bottle_fric = bottle_fric
+        self.bottle_fill_prob = bottle_fill_prob
+        self.bottle_fric_prob = bottle_fric_prob
+
+    def __repr__(self):
+        return "fill, fric, pfill, pfric: %.2f, %.2f, %.2f, %.2f" % (
+            self.bottle_fill, self.bottle_fric, self.bottle_fill_prob, self.bottle_fric_prob)
+
+
+class Environment(object):
```

Comment:
Consider if you were going to release this environment so that others could use it (to write their own methods to compare to yours, for instance), what would the easiest API be for them to use? I personally quite like the OpenAI gym spec, but whatever, as long as it's usable. (Where my definition of "environment" here also generally includes reward definition, i.e. target specification.) In your experiment_helpers you have a function that looks like it's generating a goal? AFAICT that should maybe be in Environment. And you have some object definitions in main.py - AFAICT that should also be in environment. And maybe have Environment only have a step function (and reset etc), and a different SimulationRunner or something class that knows how to pass trajectories of actions into the Environment. 

So basically Environment is responsible for its creation (objects and target) and whatever the "core" functionalities are that the env needs. And then for stuff that you expect others to be less interested in using (your trajectory running code, perhaps) you have a separate object. That also has the advantage of you can have multiple different types of trajectory runners, since it looks like you have several subtypes going on within this code (run_sim vs run_multiple_sims, or "online" vs not-online, or whatever makes the most sense).